specs/workload/Workload.md

```diff
@@ -0,0 +1,168 @@
+
+# 3. Description
+
+This API supports managing workloads focusing on the following scenarios:
+
+* Enumerate installed workloads
+* Retrieve properties of a workload
+* Remove installed packages and related resources comprising a workload
+
+## 3.1. API Structure
+
+## 3.2. Workload Definition via appxmanifest.xml
+
+A workload has the following properties:
+
+* Id -- programmatic identifier
+* DisplayName -- localizable name for user/human consumption
+* Handler -- WinRT object providing operations manipulating the workload
+
+An MSIX package can define one or more workloads. The following manifest schema is used:
+
+```xml
+<Package>
+  ...
+  <Properties>
+    <workload:Workloads>
+      <Workload Id="id" DisplayName="dn">
+        <Actions>
+          <Handler ActivatableClassId="acid">
+        </Actions>
+      </Workload>
+      ...
+    </Workload:Workloads>
+  </Properties>
+  ...
+</Package>
+```
+
+The following validations apply:
+
+* Id (id)
+  * String length: 1-255
+  * Allowed characters: [A-Za-z0-9_.]
+  * Comparison: Case-Insensitive
+  * Occurence: Unique across all workloads in a package
+* DisplayName (dn)
+  * Standard `DisplayName` definition
+* ActivatableClassId (acid)
+  * Standard `ActivatableClassId` definition
+
+# 4. Examples
+
+## 4.1. Remove all workloads
+
+Fabrikam app removing all workloads.
+
+```c#
+namespace RemoveAllWorkloads
+{
+    static class Program
+    {
+        static void Main()
+        {
+            var workloadManager = WorkloadManager.GetDefault();
+            foreach (string id in workloadManager.FindIds())
+            {
+                var workload = workloadManager.GetWorkload(id)
+                Console.Write($"Removing {workload.DisplayName}...");
+                var workloadHandler = workload.WorkloadHandler;
+                await workloadHandler.RemoveAsync();
+                Console.WriteLine($"Gone!");
+            }
+        }
+    }
+}
+```
+
+# 5. Remarks
+
+## 5.1. Platform Support
+
+This API is only available on Windows >= 10.0.19041.0 (aka 2004 aka 20H1).
+
+# 6. API Details
+
+```c# (but really MIDL3)
+namespace Microsoft.Windows.System.Workloads
+{
+    [contractversion(1)]
+    apicontract WorkloadsContract{};
+
+    [contract(WorkloadsContract, 1)]
+    enum WorkloadStatus
+    {
+        InProgress = 0,         // The request is in progress
+        CompletedSuccess = 1,   // The request completed successfully
+        CompletedFailure = 2,   // The request failed with some critical internal error.
+                                // See ExtendedError for HRESULT
+    };
+
+    [contract(WorkloadsContract, 1)]
+    runtimeclass WorkloadResult
+    {
+        WorkloadStatus Status { get; };
+        HRESULT ExtendedError { get; };
+    }
+
+    [contract(WorkloadsContract, 1)]
+    enum WorkloadProgressStatus
+    {
+        Queued = 0,             // The request is queued
+        InProgress = 1,         // The request is in progress
```

Comment:
Do these workloads support InProgressRetry?  I expect "no", and if that's supported in future work we'll be able to add it then.

Reply:
Your instincts are sharp as ever :P Nope, InProgressRetry isn't supported.

I've never even heard of such a thing and Tempo and my internet searchFu turns up no hits.

Reply:
Ahh. Nice. That's a notifications-specific thing.

In this case, workloadhandler.remove is typically package.remove, though it could be more (e.g. delete some associated files MSIX doesn't know about). Depends on the specific workload's implementation.